### PR TITLE
Fix cxa_thread_atexit() to properly register thread-local atexits

### LINKED ISSUE #10809: TLS struct destructors not called when thread is not joined
With the following on Linux (or Windows Subsytem for Linux)/Ubuntu
```
#include <stdio.h>
#include <iostream>
#if EMSCRIPTEN
include <emscripten.h>
#else
#include <chrono>
#include <thread>
#endif

struct TlsDestructionMonitor
{
        public:
                  TlsDestructionMonitor() {
                          std::cout << "ctr " << std::uintptr_t(this);
                              printf("Constructor()\n");
                                }

        void Set()
        {
                printf("set\n");
        }

  ~TlsDestructionMonitor()
  {
    printf("dest\n");
  }
};

thread_local TlsDestructionMonitor tls_destructionMonitor;
void* doWork(void *arg)
{
        printf("thread doing work\n");
        tls_destructionMonitor.Set();
        return NULL;
}

int main()
{
        pthread_t thread;
        pthread_attr_t attr;
        void * retVal;

        printf("hello\n");
        int id = pthread_create(&thread, NULL, &doWork, NULL);
        if(!id)
        {
                printf("sleeping\n");
#if EMSCRIPTEN
                emscripten_sleep(1000);
#else
                std::this_thread::sleep_for(std::chrono::milliseconds(1000));
#endif
                printf("slept\n");
                //printf("joining\n");
                //                //pthread_join(thread, &retVal);
                //                                //printf("joined\n");
                                                                                                                        //                                        }
        }
        return(0);
}
```
Compile with
`g++ 1.cpp -lpthread -pthread`
and run
`./a.out`
and you get
```
hello
sleeping
thread doing work
ctr 140391460898558Constructor()
set
dest
slept
```
The `dest` is from the destructor of the struct.  Same code with
```
em++ t.cpp -s USE_PTHREADS=1 -o t.html --emrun -s PROXY_TO_PTHREAD=1 -s ASYNCIFY=1
```
And run it in New Edge (Chrome backend) and you get
```
hello
sleeping
thread doing work
ctr 5398504Constructor()
set
slept
```
The destructor is not called.  If you replace the main thread's sleep with a join it works the same on Linux and Emscripten.

Possibly related to https://github.com/emscripten-core/emscripten/issues/9251 although that's about `exit` and ending all threads.

### PR DESCRIPTION
**Pull request by juj**
Fix cxa_thread_atexit() to properly register thread-local atexits, and not process-global atexits. (Also microcleanup code a little). Fixes #10809.

### CODE REVIEW

File: src/library_pthread.js [hunk 1]
```diff
@@ -163,11 +163,8 @@ var LibraryPThread = {
 #endif
 
     runExitHandlers: function() {
```
Comment (kripken) [inline src/library_pthread.js:165]: please add a comment here that this is run on exit on each pthread, and on the main thread (I was confused before reading enough code to realize that - I hope I understood correctly?)

Comment (kripken) [reply src/library_pthread.js:165]: Actually, how about renaming `exitHandlers` to `threadExitHandlers` as IIUC this is just for pthreads, and *not* the main thread?

File: src/library.js [hunk 2]
```diff
@@ -716,15 +714,19 @@ LibraryManager.library = {
...
   },
   __cxa_atexit: 'atexit',
 
-  // used in rust, clang when doing thread_local statics
-  __cxa_thread_atexit: 'atexit',
-  __cxa_thread_atexit_impl: 'atexit',
 #endif
 
+  // used in rust, clang when doing thread_local statics
+  __cxa_thread_atexit: 'pthread_cleanup_push',
```
[the remark this answers is not in the document]
Comment (abigailbunyan) [reply src/library.js:727]: Will `pthread_cleanup_push` be available at all if `!USE_PTHREADS`? Should these `thread_atexit`s just map to `atexit` if pthreads aren't available?

**Comment by sbc100** [comment]
I wonder if we can rely on the native libc++abi implementation.     I've started experimenting with it here and it might be possible to use this in all cases: https://github.com/emscripten-core/emscripten/pull/11040

File: src/library.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (juj) [reply src/library.js:727]: Thanks @abigailbunyan for the excellent observation. By the comment you post, `pthread_cleanup_push` handlers should only be called if thread performs an exit without returning. This PR does not change behavior around that: Both before and after this PR, `pthread_cleanup_push` handlers are called on all kinds of exits. So let's deal with that as a followup.

`pthread_cleanup_push` is available in single-threaded programs, and goes via completely different implementation in library_pthread_stub.js - which does map to atexit already.

File: src/library_pthread.js (hunk 1 above)
Comment (juj) [reply src/library_pthread.js:165]: That works, although that will be a bit redundant. These were named `PThread.exitHandlers` before, so having `PThread.threadExitHandlers` is duplicating the `thread` name. Perhaps that is worth it though.

File: src/library.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (juj) [reply src/library.js:744]: Though now I realize routing directly to atexit in singlethreaded builds generates smaller code size, so using that.

**Pull request merged**